core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java

```diff
@@ -0,0 +1,165 @@
+     */
+    public void close() throws IOException {
+      mStream.close();
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      if (!ufs.rename(mTemporaryPath, mPath)) {
+        // TODO(calvin): Log a warning if the delete fails
+        ufs.delete(mTemporaryPath, false);
+      }
+    }
+  }
+
+  /** A random id generator for worker file ids. */
+  private AtomicLong mIdGenerator;
+  /** Map of worker file ids to open under file system streams. */
+  private ConcurrentMap<Long, UnderFileSystemOutputStream> mStreams;
+
+  /**
+   * Creates a new under file system manager. Stream ids are unique to each under file system
+   * manager.
+   */
+  public UnderFileSystemManager() {
+    mIdGenerator = new AtomicLong(new Random().nextLong());
+    mStreams = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Creates a {@link UnderFileSystemOutputStream} for the given file path and adds it to the
+   * open streams map keyed with the returned worker file id.
+   *
+   * @param ufsPath the path to create in the under file system
+   * @return the worker file id which should be used to reference the open stream
+   * @throws FileAlreadyExistsException if the under file system path already exists
+   * @throws IOException if an error occurs when operating on the under file system
+   */
+  public long createFile(String ufsPath) throws FileAlreadyExistsException, IOException {
+    long workerFileId = mIdGenerator.getAndIncrement();
+    UnderFileSystemOutputStream stream =
+        new UnderFileSystemOutputStream(ufsPath, WorkerContext.getConf());
+    mStreams.put(workerFileId, stream);
+    return workerFileId;
+  }
+
+  /**
+   * Cancels the open stream associated with the given worker file id. This will close the open
+   * stream.
+   *
+   * @param workerFileId the worker id referencing an open file in the under file system
+   * @throws FileDoesNotExistException if the worker file id is not valid
+   * @throws IOException if an error occurs when operating on the under file system
+   */
+  public void cancelFile(long workerFileId) throws FileDoesNotExistException, IOException {
+    UnderFileSystemOutputStream stream = mStreams.remove(workerFileId);
+    if (stream != null) {
+      stream.cancel();
+    } else {
+      throw new FileDoesNotExistException(
+          ExceptionMessage.BAD_WORKER_FILE_ID.getMessage(workerFileId));
+    }
+  }
+
+  /**
+   * Completes an open stream associated with the given worker file id. This will close the open
+   * stream.
+   *
+   * @param workerFileId the worker id referencing an open file in the under file system
+   * @throws FileDoesNotExistException if the worker file id is not valid
+   * @throws IOException if an error occurs when operating on the under file system
+   */
+  public void completeFile(long workerFileId) throws FileDoesNotExistException, IOException {
+    UnderFileSystemOutputStream stream = mStreams.remove(workerFileId);
```

Comment:
Do you mean readding the stream to `mStreams` if a failure occurs? If an error occurs I think the stream should no longer be considered valid and retries would likely fail. Do you mean I should catch exceptions in the stream's `close` method and attempt to clean up with `delete` in case of a failure?